### PR TITLE
Add merge ts py

### PR DESCRIPTION
**Pull request by lispandfound**
See ucgmsim/merge_ts_py#1

### CODE REVIEW

**Comment by sungeunbae** [comment]
You will (perhaps) need to add installing merge_ts_loop bit to this script.
https://github.com/ucgmsim/slurm_gm_workflow/blob/master/workflow/environments/org/nesi/create_python_virtenv_mahuika.sh

**Comment by sungeunbae** [comment]
You can change the n_tasks for merge_ts here: https://github.com/ucgmsim/slurm_gm_workflow/blob/master/workflow/automation/org/nesi/config.json
While you are doing it, please also fix kisti/config.json

**Comment by lispandfound** [comment]
> You can change the n_tasks for merge_ts here: https://github.com/ucgmsim/slurm_gm_workflow/blob/master/workflow/automation/org/nesi/config.json While you are doing it, please also fix kisti/config.json

done.

**Comment by lispandfound** [comment]
> You will (perhaps) need to add installing merge_ts_loop bit to this script. https://github.com/ucgmsim/slurm_gm_workflow/blob/master/workflow/environments/org/nesi/create_python_virtenv_mahuika.sh

I believe this is unnecessary. The command `merge_ts` will be created in `$PATH` with the `pip install -e .` command executed in that script.

File: merge_ts/merge_ts_loop.pyx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    top left corner.
+    '''
+    cdef int float_size, cur_timestep, cur_component, cur_y, i, y0, local_ny, local_nx, xyts_fd, n_files
+    cdef int *component_xyts_files, *local_nxs, *local_nys, *y0s
+
+    n_files = len(_component_xyts_files)
+
+    # The lists _component_xyts_files, ... are CPython lists.
+    # If we access these inside our copying loop everything becomes very slow
+    # because we need to go to the Python interpreter. So we first copy the each
```
Comment (AndrewRidden-Harper) [inline merge_ts/merge_ts_loop.pyx:48]: small typo: delete "the" in "..copy the each..."

**Review by AndrewRidden-Harper — APPROVED** [review]

**Review by joelridden — APPROVED** [review]

File: merge_ts/merge_ts_loop.pyx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                for i in range(n_files):
+                    y0 = y0s[i]
+                    local_ny = local_nys[i]
+                    local_nx = local_nxs[i]
+                    xyts_fd = component_xyts_files[i]
+                    if y0 > cur_y:
+                        break
+                    if cur_y >= y0 + local_ny:
+                        continue
+                    # By passing None as the offset, sendfile() will read from
```
Comment (sungeunbae) [inline merge_ts/merge_ts_loop.pyx:73]: None (=NULL)?

File: merge_ts/merge_ts_loop.pyx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    # If we access these inside our copying loop everything becomes very slow
+    # because we need to go to the Python interpreter. So we first copy the each
+    # list into an equivalent C list.
+    component_xyts_files = <int *> malloc(len(_component_xyts_files) * cython.sizeof(int))
+    local_nxs = <int *> malloc(len(_local_nxs) * cython.sizeof(int))
+    local_nys = <int *> malloc(len(_local_nys) * cython.sizeof(int))
+    y0s = <int *> malloc(len(_y0s) * cython.sizeof(int))
+    # NOTE: we cannot use memcpy() here because CPython lists are not continuous
+    # chunks of memory as they are in C.
+    for i in range(len(_component_xyts_files)):
```
Comment (sungeunbae) [inline merge_ts/merge_ts_loop.pyx:56]: and this for loop has virtually no impact, I suppose?

**Review by sungeunbae — COMMENTED** [review]
Fantastic work - please just address the typo Andrew discovered. I had a few trivial questions left as comments.

File: merge_ts/merge_ts_loop.pyx (hunk 3 above)
Comment (lispandfound) [reply merge_ts/merge_ts_loop.pyx:56]: Yeah it's impact on perf is irrelevant.

File: merge_ts/merge_ts_loop.pyx (hunk 2 above)
Comment (lispandfound) [reply merge_ts/merge_ts_loop.pyx:73]: Good catch, it's leftover from when I was using `os.sendfile` instead of `sendfile` directly.

**Review by joelridden — APPROVED** [review]

**Review by claudio525 — APPROVED** [review]

**Review by sungeunbae — APPROVED** [review]

**Review by AndrewRidden-Harper — APPROVED** [review]

**Pull request merged**